### PR TITLE
Stop using defaulted functions on device

### PR DESCRIPTION
**Pull request by ibaned**
See #1470

We should also consider patching Trilinos with this.

### CODE REVIEW

**Comment by ibaned** [comment]
spot check started

**Comment by ibaned** [comment]
```
Running on machine: sems
Repository Status:  47b3c09bb1093e7244a7d5e7aeb8ddbc9c62e958 Explicitly call first/second ctors


Going to test compilers:  gcc/5.3.0 gcc/6.1.0 intel/17.0.1 clang/3.9.0 cuda/8.0.44
Testing compiler gcc/5.3.0
Testing compiler gcc/6.1.0
  Starting job gcc-5.3.0-OpenMP-release
  Starting job gcc-5.3.0-OpenMP-hwloc-release
  Starting job gcc-6.1.0-Serial-release


  PASSED gcc-6.1.0-Serial-release
Testing compiler intel/17.0.1
  Starting job gcc-6.1.0-Serial-hwloc-release
  PASSED gcc-5.3.0-OpenMP-release
  Starting job intel-17.0.1-OpenMP-release
  PASSED gcc-5.3.0-OpenMP-hwloc-release
Testing compiler clang/3.9.0
  Starting job intel-17.0.1-OpenMP-hwloc-release
  PASSED gcc-6.1.0-Serial-hwloc-release
  Starting job clang-3.9.0-Pthread_Serial-release
  PASSED clang-3.9.0-Pthread_Serial-release
Testing compiler cuda/8.0.44
  Starting job clang-3.9.0-Pthread_Serial-hwloc-release
  PASSED intel-17.0.1-OpenMP-release
  PASSED intel-17.0.1-OpenMP-hwloc-release
  PASSED clang-3.9.0-Pthread_Serial-hwloc-release
  Starting job cuda-8.0.44-Cuda_OpenMP-release
  PASSED cuda-8.0.44-Cuda_OpenMP-release
#######################################################
PASSED TESTS
#######################################################
clang-3.9.0-Pthread_Serial-hwloc-release build_time=196 run_time=92
clang-3.9.0-Pthread_Serial-release build_time=199 run_time=245
cuda-8.0.44-Cuda_OpenMP-release build_time=497 run_time=560
gcc-5.3.0-OpenMP-hwloc-release build_time=267 run_time=373
gcc-5.3.0-OpenMP-release build_time=264 run_time=348
gcc-6.1.0-Serial-hwloc-release build_time=192 run_time=126
gcc-6.1.0-Serial-release build_time=211 run_time=399
intel-17.0.1-OpenMP-hwloc-release build_time=554 run_time=231
intel-17.0.1-OpenMP-release build_time=564 run_time=227
```

File: core/src/Kokkos_MemoryPool.hpp [hunk 1]
```diff
@@ -258,10 +258,56 @@ class MemoryPool {
...
+    m_tracker = rhs.m_tracker;
+    m_sb_state_array = rhs.m_sb_state_array;
+    m_sb_state_size = rhs.m_sb_state_size;
+    m_sb_size_lg2 = rhs.m_sb_size_lg2;
+    m_max_block_size_lg2 = rhs.m_max_block_size_lg2;
+    m_min_block_size_lg2 = rhs.m_min_block_size_lg2;
+    m_sb_count = rhs.m_sb_count;
+    m_hint_offset = rhs.m_hint_offset;
+    m_data_offset = rhs.m_data_offset;
+    return *this;
```
Comment (dalg24) [inline core/src/Kokkos_MemoryPool.hpp]:
```C++
if ( this != other) {
   ...
}
return *this;
```

Comment (mhoemmen) [reply core/src/Kokkos_MemoryPool.hpp]: @dalg24 Good point, though I think the self-assignments may be harmless in this case.

Comment (ibaned) [reply core/src/Kokkos_MemoryPool.hpp]: @dalg24 yes thats a useful optimization in some cases, but I think the current use cases for MemoryPool will never hit this case.

**Comment by crtrott** [comment]
I discussed this with Dan, since this is basically a bug in CUDA 9 (and the bug is actually the warnings because it apparently still works if you simply use KOKKOS_INLINE_FUNCTION to mark the defaulted functions) we add a KOKKOS_BUG_CUDA9_DEFAULTED_FUNCTIONS_WORKAROUND and have the "define all functions explicitly code" only on for that, otherwise still use the defaulted functions. We nevertheless remove the _DEFAULTED macros since they are not actually needed in that case, Dan is going to apply this fix explicitly to Trilinos, and we do a bug fix release of Kokkos next week.

**Comment by ibaned** [comment]
Still passes spot check:
```
Running on machine: sems
Repository Status:  f91c031a391831ab2727707e42d3d9b4f00bfe60 One last touch-up to Pair ctor


Going to test compilers:  gcc/5.3.0 gcc/6.1.0 intel/17.0.1 clang/3.9.0 cuda/8.0.44
Testing compiler gcc/5.3.0
Testing compiler gcc/6.1.0
  Starting job gcc-5.3.0-OpenMP-release
  Starting job gcc-5.3.0-OpenMP-hwloc-release
  Starting job gcc-6.1.0-Serial-release
  PASSED gcc-6.1.0-Serial-release
Testing compiler intel/17.0.1
  Starting job gcc-6.1.0-Serial-hwloc-release
  PASSED gcc-5.3.0-OpenMP-release
  Starting job intel-17.0.1-OpenMP-release
  PASSED gcc-5.3.0-OpenMP-hwloc-release
Testing compiler clang/3.9.0
  Starting job intel-17.0.1-OpenMP-hwloc-release
  PASSED gcc-6.1.0-Serial-hwloc-release
  Starting job clang-3.9.0-Pthread_Serial-release
  PASSED clang-3.9.0-Pthread_Serial-release
Testing compiler cuda/8.0.44
  Starting job clang-3.9.0-Pthread_Serial-hwloc-release
  PASSED intel-17.0.1-OpenMP-release
  PASSED intel-17.0.1-OpenMP-hwloc-release
  PASSED clang-3.9.0-Pthread_Serial-hwloc-release
  Starting job cuda-8.0.44-Cuda_OpenMP-release
  PASSED cuda-8.0.44-Cuda_OpenMP-release
#######################################################
PASSED TESTS
#######################################################
clang-3.9.0-Pthread_Serial-hwloc-release build_time=215 run_time=106
clang-3.9.0-Pthread_Serial-release build_time=328 run_time=230
cuda-8.0.44-Cuda_OpenMP-release build_time=522 run_time=414
gcc-5.3.0-OpenMP-hwloc-release build_time=355 run_time=294
gcc-5.3.0-OpenMP-release build_time=348 run_time=297
gcc-6.1.0-Serial-hwloc-release build_time=379 run_time=161
gcc-6.1.0-Serial-release build_time=284 run_time=232
intel-17.0.1-OpenMP-hwloc-release build_time=840 run_time=186
intel-17.0.1-OpenMP-release build_time=832 run_time=193
```

**Pull request merged**